x/vesting/types/clawback_vesting_account.go

```diff
@@ -140,90 +140,63 @@ func (va ClawbackVestingAccount) GetLockedOnly(blockTime time.Time) sdk.Coins {
 	return va.OriginalVesting.Sub(va.GetUnlockedOnly(blockTime))
 }
 
-// GetVestedOnly returns the vesting schedule and blockTime.
+// GetVestedOnly returns the vesting schedule at blockTime.
 func (va ClawbackVestingAccount) GetVestedOnly(blockTime time.Time) sdk.Coins {
 	return ReadSchedule(va.GetStartTime(), va.EndTime, va.VestingPeriods, va.OriginalVesting, blockTime.Unix())
 }
 
-// GetUNvestedOnly returns the unvesting schedule and blockTime.
+// GetUNvestedOnly returns the unvesting schedule at blockTime.
 func (va ClawbackVestingAccount) GetUnvestedOnly(blockTime time.Time) sdk.Coins {
-	return va.OriginalVesting.Sub(va.GetVestedOnly(blockTime))
+	totalUnvested := va.OriginalVesting.Sub(va.GetVestedOnly(blockTime))
+	if totalUnvested == nil {
+		totalUnvested = sdk.Coins{}
+	}
+	return totalUnvested
+}
+
+// GetPassedPeriodCount returns the amount of passed periods at blockTime.
+func (va ClawbackVestingAccount) GetPassedPeriodCount(blockTime time.Time) int {
+	return ReadPastPeriodCount(va.GetStartTime(), va.EndTime, va.VestingPeriods, blockTime.Unix())
 }
 
-// ComputeClawback returns an account with all future vesting events removed,
-// plus the total sum of these events. When removing the future vesting events,
-// the lockup schedule will also have to be capped to keep the total sums the same.
-// (But future unlocking events might be preserved if they unlock currently vested coins.)
-// If the amount returned is zero, then the returned account should be unchanged.
-// Does not adjust DelegatedVesting
+// ComputeClawback returns an account with all future vesting events removed and
+// the clawback amount (total sum of these events). Future unlocking events are
+// preserved and update in case unlocked vested coins remain after clawback.
 func (va ClawbackVestingAccount) ComputeClawback(
 	clawbackTime int64,
 ) (ClawbackVestingAccount, sdk.Coins) {
-	// Compute the truncated vesting schedule and amounts.
-	// Work with the schedule as the primary data and recompute derived fields, e.g. OriginalVesting.
-	t := va.GetStartTime()
-	totalVested := sdk.NewCoins()
-	totalUnvested := sdk.NewCoins()
-	unvestedIdx := 0
-
-	for i, period := range va.VestingPeriods {
-		t += period.Length
-		// tie in time goes to clawback
-		if t < clawbackTime {
-			totalVested = totalVested.Add(period.Amount...)
-			unvestedIdx = i + 1
-		} else {
-			totalUnvested = totalUnvested.Add(period.Amount...)
-		}
-	}
+	totalVested := va.GetVestedOnly(time.Unix(clawbackTime, 0))
+	totalUnvested := va.GetUnvestedOnly(time.Unix(clawbackTime, 0))
 
-	newVestingPeriods := va.VestingPeriods[:unvestedIdx]
+	// Remove all unvested periods from the schdule
+	passedPeriodID := va.GetPassedPeriodCount(time.Unix(clawbackTime, 0))
+	newVestingPeriods := va.VestingPeriods[:passedPeriodID]
 
-	// To cap the unlocking schedule to the new total vested, conjunct with a limiting schedule
+	newVestingEnd := va.GetStartTime()
+	for _, period := range va.VestingPeriods {
```

Comment:
I see `newVestingEnd` is used for the new `va.EndTime` value. I wonder why not `for _, period := range newVestingPeriods {`. 
In the case of a clawback, I would think that the end of a vesting account can be at the time of clawback (if unlocked) or when the unlock period finishes. But I do not know if this `EndTime` has any implications, so this is more of a question.

Reply:
@ramacarlucho

Reply:
Good catch,